# Домашнее задание к лекции "Базовые понятия статистики"

## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

## Дополнительная часть (необязательная)

Выполнить задания 1-3 для всего набора данных.

### Решение

In [544]:
import pandas as pd
import numpy as np

# Задание 1. Базовое изучение
## По результатам анализа присутствуют следующие метрики

## Числовые: 
rectal temperature, pulse, respiratory rate, packed cell volume, total protein

## Категориальные: 
surgery, Age, Hospital Number, temperature of extremities, peripheral pulse, mucous membranes, capillary refill time, pain, peristalsis, abdominal distension, nasogastric tube, nasogastric reflux, nasogastric reflux PH, rectal examination, abdomen, abdomcentesis total protein, abdominocentesis appearance, outcome, surgical lesion, type of lesion1,type of lesion2, type of lesion3, cp_data


In [545]:
# прочитаем файл с данными и назовём столбцы согласно документации на них
df = pd.read_csv('horse_data.csv', na_values='?')
df.columns = ['surgery','Age','Hospital Number','rectal temperature','pulse','respiratory rate',
                   'temperature of extremities','peripheral pulse','mucous membranes','capillary refill time',
                   'pain','peristalsis','abdominal distension','nasogastric tube','nasogastric reflux','nasogastric reflux PH',
                   'rectal examination','abdomen','packed cell volume','total protein','abdominocentesis appearance',
                   'abdomcentesis total protein','outcome','surgical lesion','type of lesion1','type of lesion2',
                   'type of lesion3','cp_data']
# выберем 8 столбцов, с которыми будем работать
df = df.iloc[:, 0:8]
df

surgery  Age  Hospital Number  rectal temperature  pulse  \
0        1.0    1           534817                39.2   88.0   
1        2.0    1           530334                38.3   40.0   
2        1.0    9          5290409                39.1  164.0   
3        2.0    1           530255                37.3  104.0   
4        2.0    1           528355                 NaN    NaN   
..       ...  ...              ...                 ...    ...   
294      1.0    1           533886                 NaN  120.0   
295      2.0    1           527702                37.2   72.0   
296      1.0    1           529386                37.5   72.0   
297      1.0    1           530612                36.5  100.0   
298      1.0    1           534618                37.2   40.0   

     respiratory rate  temperature of extremities  peripheral pulse  
0                20.0                         NaN               NaN  
1                24.0                         1.0               1.0  
2                84.0                         4.0               1.0  
3                35.0                         NaN               NaN  
4                 NaN                         2.0               1.0  
..                ...                         ...               ...  
294              70.0                         4.0               NaN  
295              24.0                         3.0               2.0  
296              30.0                         4.0               3.0  
297              24.0                         3.0               3.0  
298              20.0                         NaN               NaN  

[299 rows x 8 columns]

In [546]:
# отображение некоторых статистик максимального и минимального значений, размаха, медиана и моды
print('Maximums\n{}\n\n Minimums\n{}\n\n Differences\n{}\n\n Medians\n{}\n\n Mode\n{}\n\n'.format(df.max(), df.min(), df.max()-df.min(), df.median(), df.mode()))


Maximums
surgery                             2.0
Age                                 9.0
Hospital Number               5305629.0
rectal temperature                 40.8
pulse                             184.0
respiratory rate                   96.0
temperature of extremities          4.0
peripheral pulse                    4.0
dtype: float64

 Minimums
surgery                            1.0
Age                                1.0
Hospital Number               518476.0
rectal temperature                35.4
pulse                             30.0
respiratory rate                   8.0
temperature of extremities         1.0
peripheral pulse                   1.0
dtype: float64

 Differences
surgery                             1.0
Age                                 8.0
Hospital Number               4787153.0
rectal temperature                  5.4
pulse                             154.0
respiratory rate                   88.0
temperature of extremities          3.0
peripheral pulse        

In [547]:
# небольшое описание датафрейма по каждому столбцу и статистикам
df.describe()

surgery        Age  Hospital Number  rectal temperature       pulse  \
count  298.000000  299.00000     2.990000e+02          239.000000  275.000000   
mean     1.395973    1.64214     1.087748e+06           38.166527   71.934545   
std      0.489881    2.17730     1.532026e+06            0.733508   28.680522   
min      1.000000    1.00000     5.184760e+05           35.400000   30.000000   
25%      1.000000    1.00000     5.289040e+05           37.800000   48.000000   
50%      1.000000    1.00000     5.303100e+05           38.200000   64.000000   
75%      2.000000    1.00000     5.347360e+05           38.500000   88.000000   
max      2.000000    9.00000     5.305629e+06           40.800000  184.000000   

       respiratory rate  temperature of extremities  peripheral pulse  
count        241.000000                  243.000000        230.000000  
mean          30.427386                    2.345679          2.013043  
std           17.678256                    1.046369          1.042672  
min            8.000000                    1.000000          1.000000  
25%           18.000000                    1.000000          1.000000  
50%           24.000000                    3.000000          1.500000  
75%           36.000000                    3.000000          3.000000  
max           96.000000                    4.000000          4.000000

In [548]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     298 non-null    float64
 1   Age                         299 non-null    int64  
 2   Hospital Number             299 non-null    int64  
 3   rectal temperature          239 non-null    float64
 4   pulse                       275 non-null    float64
 5   respiratory rate            241 non-null    float64
 6   temperature of extremities  243 non-null    float64
 7   peripheral pulse            230 non-null    float64
dtypes: float64(6), int64(2)
memory usage: 18.8 KB


In [549]:
# первый и третий квартиль (25-й и 75-й процентиль)
df.quantile([0.25, 0.75])

surgery  Age  Hospital Number  rectal temperature  pulse  \
0.25      1.0  1.0         528904.0                37.8   48.0   
0.75      2.0  1.0         534736.0                38.5   88.0   

      respiratory rate  temperature of extremities  peripheral pulse  
0.25              18.0                         1.0               1.0  
0.75              36.0                         3.0               3.0

In [550]:
# СКО - среднеквадратичное отклонение
df.std()

surgery                       4.898813e-01
Age                           2.177300e+00
Hospital Number               1.532026e+06
rectal temperature            7.335084e-01
pulse                         2.868052e+01
respiratory rate              1.767826e+01
temperature of extremities    1.046369e+00
peripheral pulse              1.042672e+00
dtype: float64

In [551]:
# дисперсия
df.var()

surgery                       2.399837e-01
Age                           4.740634e+00
Hospital Number               2.347104e+12
rectal temperature            5.380345e-01
pulse                         8.225723e+02
respiratory rate              3.125207e+02
temperature of extremities    1.094888e+00
peripheral pulse              1.087165e+00
dtype: float64

### Небольшое описание полученных данных

Надо отметить, что в данных существуют как пропуски данных, так и выбросы. Вместо пропусков фигурировали знаки вопросов ? - чтобы возможно было посчитать статистики, все знаки вопроса были заменены на пустоты. Сразу понятно, что некоторые статистики бессмысленны для столбца с категориальной величиной, к примеру, Hospital number, однако с другими категориальными величинами Age, мы можем по среднему сказать, каких больше, а для количественных данных всё понятно. 

Фигурирование  пропусков и выбросов оказывает влияние на данные, поэтому необходимо решить, откуда появляются такие "ломаные" данные, что с ними делать, как считать. См. следующие задания... 

# Задание 2. Работа с выбросами

In [552]:
# найдём выбросы в каждом столбце
dict_border_outliers = {}
for column in df:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_out = Q1 - (1.5 * IQR)
#так как здесь не может быть отрицательной границы, то ставим минимальное значение, обозначающие наши пропуски (!просто нули здесь невозможны)
    if lower_out < 0: lower_out = 0
    higher_out = Q3 + (1.5 * IQR)
    dict_border_outliers.setdefault(column, [lower_out, higher_out])
dict_border_outliers

{'surgery': [0, 3.5],
 'Age': [1.0, 1.0],
 'Hospital Number': [520156.0, 543484.0],
 'rectal temperature': [36.74999999999999, 39.550000000000004],
 'pulse': [0, 148.0],
 'respiratory rate': [0, 63.0],
 'temperature of extremities': [0, 6.0],
 'peripheral pulse': [0, 6.0]}

In [553]:
# выведем на экран отдельно исследуемые столбцы без выбросов, очерчивая границу посчитанную выше
# проанализируем ситуацию
for column in df:
    outliers = df[df[column].between(dict_border_outliers[column][0], dict_border_outliers[column][1],inclusive=True)].sort_values(column)
    print(column,'\n', outliers[column],'\n')

surgery 
 0      1.0
168    1.0
171    1.0
172    1.0
174    1.0
      ... 
194    2.0
195    2.0
196    2.0
190    2.0
149    2.0
Name: surgery, Length: 298, dtype: float64 

Age 
 0      1
188    1
189    1
191    1
192    1
      ..
100    1
99     1
98     1
105    1
298    1
Name: Age, Length: 275, dtype: int64 

Hospital Number 
 196    521399
21     521681
140    522979
176    523190
16     526639
        ...  
99     535364
88     535381
223    535392
43     535407
28     535415
Name: Hospital Number, Length: 263, dtype: int64 

rectal temperature 
 112    36.8
276    36.9
291    37.0
69     37.0
262    37.1
       ... 
119    39.4
286    39.5
161    39.5
230    39.5
116    39.5
Name: rectal temperature, Length: 225, dtype: float64 

pulse 
 129     30.0
231     30.0
275     36.0
202     36.0
241     36.0
       ...  
147    132.0
134    136.0
43     140.0
102    140.0
38     146.0
Name: pulse, Length: 270, dtype: float64 

respiratory rate 
 189     8.0
202     9.0
139     9.0

In [554]:
# выведем на экран отдельно исследуемые столбцы без выбросов, очерчивая границу посчитанную выше
# проанализируем ситуацию
for column in df:
    df = df[df[column].between(dict_border_outliers[column][0], dict_border_outliers[column][1], inclusive=True)]
df

surgery  Age  Hospital Number  rectal temperature  pulse  \
1        2.0    1           530334                38.3   40.0   
5        1.0    1           526802                37.9   48.0   
9        1.0    1           528548                38.1   66.0   
11       1.0    1           528031                37.2   42.0   
13       1.0    1           534917                38.2   76.0   
..       ...  ...              ...                 ...    ...   
290      1.0    1           528890                38.9   80.0   
291      1.0    1           530034                37.0   66.0   
293      2.0    1           533902                38.5   40.0   
295      2.0    1           527702                37.2   72.0   
296      1.0    1           529386                37.5   72.0   

     respiratory rate  temperature of extremities  peripheral pulse  
1                24.0                         1.0               1.0  
5                16.0                         1.0               1.0  
9                12.0                         3.0               3.0  
11               12.0                         2.0               1.0  
13               28.0                         3.0               1.0  
..                ...                         ...               ...  
290              44.0                         3.0               3.0  
291              20.0                         1.0               3.0  
293              16.0                         1.0               1.0  
295              24.0                         3.0               2.0  
296              30.0                         4.0               3.0  

[120 rows x 8 columns]

In [555]:
df.head()

surgery  Age  Hospital Number  rectal temperature  pulse  \
1       2.0    1           530334                38.3   40.0   
5       1.0    1           526802                37.9   48.0   
9       1.0    1           528548                38.1   66.0   
11      1.0    1           528031                37.2   42.0   
13      1.0    1           534917                38.2   76.0   

    respiratory rate  temperature of extremities  peripheral pulse  
1               24.0                         1.0               1.0  
5               16.0                         1.0               1.0  
9               12.0                         3.0               3.0  
11              12.0                         2.0               1.0  
13              28.0                         3.0               1.0

In [556]:
# не смотреть - сделано для проверки выбросов
df[df['Age'] > 0 ]

surgery  Age  Hospital Number  rectal temperature  pulse  \
1        2.0    1           530334                38.3   40.0   
5        1.0    1           526802                37.9   48.0   
9        1.0    1           528548                38.1   66.0   
11       1.0    1           528031                37.2   42.0   
13       1.0    1           534917                38.2   76.0   
..       ...  ...              ...                 ...    ...   
290      1.0    1           528890                38.9   80.0   
291      1.0    1           530034                37.0   66.0   
293      2.0    1           533902                38.5   40.0   
295      2.0    1           527702                37.2   72.0   
296      1.0    1           529386                37.5   72.0   

     respiratory rate  temperature of extremities  peripheral pulse  
1                24.0                         1.0               1.0  
5                16.0                         1.0               1.0  
9                12.0                         3.0               3.0  
11               12.0                         2.0               1.0  
13               28.0                         3.0               1.0  
..                ...                         ...               ...  
290              44.0                         3.0               3.0  
291              20.0                         1.0               3.0  
293              16.0                         1.0               1.0  
295              24.0                         3.0               2.0  
296              30.0                         4.0               3.0  

[120 rows x 8 columns]

## Выбросы и их причины 
У нас всего 299 строк значений, разберём те, у кого отсеились выбросы. Итого, осталось строк у следующих столбцов: 
Surgery - 298; Age - 275, Hospital Number - 263, rectal temperature - 225, pulse - 270, respiratory rate - 224; temperature of extremities - 243; peripheral pulse - 230

Причины:

Age - скорее ошибки ввода данных(человеческий фактор), либо ошибки обработки данных -попадение случайно, т.к. к примеру число 9 не обозначает ничего в описании столбца;

Hospital Number - здесь так же ошибки ввода данных(человеческий фактор)/опечатка, случай введения на одну цифру больше, чем должно быть, однако в теории, возможно, всё-таки id лошадки, но слишком сильно выбивается номер на порядок больше остальных, поэтому я бы не учитывал эти данные или проконсультировался со сборщиком данных=); 

rectal temperature - из описания "этот параметр обычно изменяется по мере развития проблемы, например, может начинаться нормально, затем повышаться из-за поражения, переходя обратно в нормальный диапазон, когда лошадь впадает в шок" - трудно сказать, что здесь выброс, а что реальные значения; 

pulse - по описанию, нормальный пульс может быть у спортивных лошадей 20-25 (30-40 взрослые), а с болезненными поражениями или в шоке может быть повышенный пульс, сильных выбросов нет, которые есть, это только не совершенные рамки выбросов, однако есть и значение 180 скорее ошибка измерителя; 

respiratory rate - как написано в описании столбца, полезность сомнительна из-за больших колебаний, так что выход за рамки лишь условное ограничение, как и условны сами ограничения

Решение по выбросам

Можно использовать условно принятый интервал через IQR и Q1, Q2, исключив все выбросы, которые выходят за рамка интервала, остальные значения выходящие за рамки можно либо принять за погрешность нашего метода расчёта, либо расширять рамки до адекватных возможных значений

Итого, после чистки от строк с выбросами, уменьшилось количество данных на 60%, что не супер (было 298, стало 120) Зато они все чистенькие=)

# Задание 3. Работа с пропусками

In [573]:
# прочитаем файл с данными и назовём столбцы согласно документации на них
df = pd.read_csv('horse_data.csv', na_values='?')
df.columns = ['surgery','Age','Hospital Number','rectal temperature','pulse','respiratory rate',
                   'temperature of extremities','peripheral pulse','mucous membranes','capillary refill time',
                   'pain','peristalsis','abdominal distension','nasogastric tube','nasogastric reflux','nasogastric reflux PH',
                   'rectal examination','abdomen','packed cell volume','total protein','abdominocentesis appearance',
                   'abdomcentesis total protein','outcome','surgical lesion','type of lesion1','type of lesion2',
                   'type of lesion3','cp_data']
# выберем 8 столбцов, с которыми будем работать
df = df.iloc[:, 0:8]
df

surgery  Age  Hospital Number  rectal temperature  pulse  \
0        1.0    1           534817                39.2   88.0   
1        2.0    1           530334                38.3   40.0   
2        1.0    9          5290409                39.1  164.0   
3        2.0    1           530255                37.3  104.0   
4        2.0    1           528355                 NaN    NaN   
..       ...  ...              ...                 ...    ...   
294      1.0    1           533886                 NaN  120.0   
295      2.0    1           527702                37.2   72.0   
296      1.0    1           529386                37.5   72.0   
297      1.0    1           530612                36.5  100.0   
298      1.0    1           534618                37.2   40.0   

     respiratory rate  temperature of extremities  peripheral pulse  
0                20.0                         NaN               NaN  
1                24.0                         1.0               1.0  
2                84.0                         4.0               1.0  
3                35.0                         NaN               NaN  
4                 NaN                         2.0               1.0  
..                ...                         ...               ...  
294              70.0                         4.0               NaN  
295              24.0                         3.0               2.0  
296              30.0                         4.0               3.0  
297              24.0                         3.0               3.0  
298              20.0                         NaN               NaN  

[299 rows x 8 columns]

In [574]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     298 non-null    float64
 1   Age                         299 non-null    int64  
 2   Hospital Number             299 non-null    int64  
 3   rectal temperature          239 non-null    float64
 4   pulse                       275 non-null    float64
 5   respiratory rate            241 non-null    float64
 6   temperature of extremities  243 non-null    float64
 7   peripheral pulse            230 non-null    float64
dtypes: float64(6), int64(2)
memory usage: 18.8 KB


In [575]:
print(df['peripheral pulse'].mean())
print(df['peripheral pulse'].median())
print(df['peripheral pulse'].mode()[0])

print(df['peripheral pulse'].std())
print(df['peripheral pulse'].var())

2.0130434782608697
1.5
1.0
1.0426722252368779
1.0871653692804226


In [576]:
# проверим сколько пропусков здесь
for col in df:
    df_missing = df[col].isna().mean()
    print(f'{col} - {df_missing :.1%}')

surgery - 0.3%
Age - 0.0%
Hospital Number - 0.0%
rectal temperature - 20.1%
pulse - 8.0%
respiratory rate - 19.4%
temperature of extremities - 18.7%
peripheral pulse - 23.1%


## Анализ пропусков
По результатам можем увидеть, что наибольшее число пропусков у колонок(ТОП-4): peripheral pulse, rectal temperature, respiratory rate, temperature of extremities - свыше 18%. Поработаем с пропусками. Посмотрев ниже, можно убедиться в том, что если избавиться путём удаления всех строк с пропусками, то получим потери данных около 48% - это естественно никуда не гадиться. Попробуем  оставить строки, где 7 из 8 параметров заполнено, затем заполнить количественные типы данных - медианой, а качественные - модой и посмотрим на результаты. Однако будем менять не просто модой и медианой по всем значениям, а в идеале с учётом возраста и хирургического вмешательства (т.е. Age, surgery), однако мы знаем, что у нас есть выбросы в данных о возрасте, а в surgery нет, поэтому будем использовать только хирургическое вмешательство

In [577]:
df.dropna().info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 150 entries, 1 to 297
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     150 non-null    float64
 1   Age                         150 non-null    int64  
 2   Hospital Number             150 non-null    int64  
 3   rectal temperature          150 non-null    float64
 4   pulse                       150 non-null    float64
 5   respiratory rate            150 non-null    float64
 6   temperature of extremities  150 non-null    float64
 7   peripheral pulse            150 non-null    float64
dtypes: float64(6), int64(2)
memory usage: 10.5 KB


In [578]:
print(df.dropna()['peripheral pulse'].mean())
print(df.dropna()['peripheral pulse'].median())
print(df.dropna()['peripheral pulse'].mode()[0])

print(df.dropna()['peripheral pulse'].std())
print(df.dropna()['peripheral pulse'].var())


1.86
1.0
1.0
1.0035507430916137
1.00711409395973


In [579]:
df.dropna(thresh=7).info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 222 entries, 1 to 297
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     221 non-null    float64
 1   Age                         222 non-null    int64  
 2   Hospital Number             222 non-null    int64  
 3   rectal temperature          201 non-null    float64
 4   pulse                       219 non-null    float64
 5   respiratory rate            202 non-null    float64
 6   temperature of extremities  213 non-null    float64
 7   peripheral pulse            204 non-null    float64
dtypes: float64(6), int64(2)
memory usage: 15.6 KB


In [580]:
df = df.dropna(thresh=7)
df

surgery  Age  Hospital Number  rectal temperature  pulse  \
1        2.0    1           530334                38.3   40.0   
2        1.0    9          5290409                39.1  164.0   
5        1.0    1           526802                37.9   48.0   
7        2.0    1           530051                 NaN   80.0   
9        1.0    1           528548                38.1   66.0   
..       ...  ...              ...                 ...    ...   
292      1.0    1           534004                 NaN   78.0   
293      2.0    1           533902                38.5   40.0   
295      2.0    1           527702                37.2   72.0   
296      1.0    1           529386                37.5   72.0   
297      1.0    1           530612                36.5  100.0   

     respiratory rate  temperature of extremities  peripheral pulse  
1                24.0                         1.0               1.0  
2                84.0                         4.0               1.0  
5                16.0                         1.0               1.0  
7                36.0                         3.0               4.0  
9                12.0                         3.0               3.0  
..                ...                         ...               ...  
292              24.0                         3.0               3.0  
293              16.0                         1.0               1.0  
295              24.0                         3.0               2.0  
296              30.0                         4.0               3.0  
297              24.0                         3.0               3.0  

[222 rows x 8 columns]

In [581]:
quality_data = ['surgery', 'Age', 'Hospital Number', 'temperature of extremities', 'peripheral pulse']
quantitative_data = ['rectal temperature', 'pulse', 'respiratory rate']

In [625]:
dg = df
dg

surgery  Age  Hospital Number  rectal temperature  pulse  \
1        2.0    1           530334                38.3   40.0   
2        1.0    9          5290409                39.1  164.0   
5        1.0    1           526802                37.9   48.0   
7        2.0    1           530051                 NaN   80.0   
9        1.0    1           528548                38.1   66.0   
..       ...  ...              ...                 ...    ...   
292      1.0    1           534004                 NaN   78.0   
293      2.0    1           533902                38.5   40.0   
295      2.0    1           527702                37.2   72.0   
296      1.0    1           529386                37.5   72.0   
297      1.0    1           530612                36.5  100.0   

     respiratory rate  temperature of extremities  peripheral pulse  
1                24.0                         1.0               1.0  
2                84.0                         4.0               1.0  
5                16.0                         1.0               1.0  
7                36.0                         3.0               4.0  
9                12.0                         3.0               3.0  
..                ...                         ...               ...  
292              24.0                         3.0               3.0  
293              16.0                         1.0               1.0  
295              24.0                         3.0               2.0  
296              30.0                         4.0               3.0  
297              24.0                         3.0               3.0  

[222 rows x 8 columns]

In [629]:
#распределение замен пропусков: количественный тип данных - медиана, качественный - мода
for col in dg.columns:
    if col in quality_data: 
        dg = dg[col].fillna(df.groupby(['surgery'])[col].mode()[0], inplace = True)
    elif col in quantitative_data: 
        dg = dg[col].fillna(df.groupby(['surgery'])[col].transform('median'), inplace = True)
dg

AttributeError: 'NoneType' object has no attribute 'columns'

In [630]:
# # проверим сколько пропусков здесь
# for col in dg:
#     dg_missing = dg[col].isna().mean()
#     print(f'{col} - {dg_missing :.1%}')

TypeError: 'NoneType' object is not iterable